Use wxOverlay to show sash feedback in wxAUI and in wxSplitterWindow

AliKet opened the pull request:
See #24177

Tested under: `wxMSW`, `wxGTK3` (**Wayland** & **X11**) and `wxQt`

Not Tested under `wxMac` though!

AliKet committed:
Remove unnecessary override from wxOverlayImpl under wxMSW

The base class version already returns true by default.

AliKet commented:
May conflict with #24166

AliKet commented:
I just tested this under `wxGTK2` and didn't work unfortunately.

AliKet commented:
@vadz @paulcor  If I apply this patch:
```
diff --git a/src/gtk/dcclient.cpp b/src/gtk/dcclient.cpp
index 69478afc92..c2ed3f7afa 100644
--- a/src/gtk/dcclient.cpp
+++ b/src/gtk/dcclient.cpp
@@ -452,6 +452,11 @@ void wxWindowDCImpl::SetUpDC( bool isMemDC )
     gdk_gc_set_clip_rectangle( m_brushGC, nullptr );
     gdk_gc_set_clip_rectangle( m_textGC, nullptr );
     gdk_gc_set_clip_rectangle( m_bgGC, nullptr );
+
+    gdk_gc_set_subwindow( m_penGC, GDK_INCLUDE_INFERIORS );
+    gdk_gc_set_subwindow( m_brushGC, GDK_INCLUDE_INFERIORS );
+    gdk_gc_set_subwindow( m_textGC, GDK_INCLUDE_INFERIORS );
+    gdk_gc_set_subwindow( m_bgGC, GDK_INCLUDE_INFERIORS );
 }
 
 void wxWindowDCImpl::DoGetSize( int* width, int* height ) const
```
then this PR works correctly under wxGTK2.

vadz commented:
Sorry, I've just realized that I perhaps shouldn't have merged #24166 before this one. But I can resolve the conflicts, however I'd like to understand the situation with wxGTK2: can we use `wxOverlay` for it or not, finally? I'd like to minimize time spent on it, but AFAIK we don't plan to drop support for it yet, so we still need this to work in at least some way there...

Commit:
Use wxOverlay to show sash feedback in non-live resize mode in wxAUI

Make non-live resizing possible on all supported platforms thanks to wxOverlay.
Previously, live resizing was the only supported mode under these platforms:
wxMac, wxQt and wxGTK3 under Wayland.

Commit:
Use wxOverlay to show sash feedback in non-live resize mode in wxSplitterWindow

Make non-live resizing possible on all supported platforms thanks to wxOverlay.
Previously, live resizing was the only supported mode under these platforms:
wxMac, wxQt and wxGTK3 under Wayland.

Also remove no longer needed function AlwaysUsesLiveUpdate and its documentation.

Commit:
Make wxAUI and wxSplitterWindow work properly with wxOverlay under wxGTK2

This Workaround is needed because the DC used to draw on the wxOverlay clips
children by default under wxGTK2. This means that the sash will always appear
behind the sub-windows. making wxOverlay worthless under this platforms.

Notice that calling gdk_gc_set_subwindow() unconditionally in Destroy() is cheap
when DontClipSubWindows() has never been called.

AliKet commented:
> Sorry, I've just realized that I perhaps shouldn't have merged #24166 before this one. But I can resolve the conflicts, however I'd like to understand the situation with wxGTK2: can we use `wxOverlay` for it or not, finally? I'd like to minimize time spent on it, but AFAIK we don't plan to drop support for it yet, so we still need this to work in at least some way there...

`wxAUI` and `wxSplitterWindow` now work correctly under all supported platforms (can not test under **wxMac** !!)

AliKet commented:
I also managed to get transparent hint to work with `wxOverlay` (generic and native), so `wxPseudoTransparentFrame` (with all its complications and limitation) is no longer needed.
I just want to know if it's ok to push the changes (depend on the changes done here) here or should I wait until this is approved and merged?

as always I need someone else to test my changes under **wxOSX**. TIA!

vadz commented on src/msw/overlay.cpp:
```diff
@@ -100,8 +99,16 @@ void wxOverlayImpl::Init(wxDC* dc, int , int , int , int )
 
-    m_rect.SetSize(m_window->GetClientSize());
-    m_rect.SetPosition(m_window->GetScreenPosition());
+    if ( wxDynamicCast(dc, wxClientDC) )
+    {
+        m_rect = m_window->GetClientRect();
+        // The rectangle should be in screen coordinates
+        m_window->ClientToScreen(&m_rect.x, &m_rect.y);
+    }
+    else // wxWindowDC
+    {
+        m_rect = m_window->GetScreenRect();
+    }
```
Why can't we always just use `GetScreenRect()` here? Is it just because we really need client size for `wxClientDC` but full window size otherwise? If so, we could still use something like

```c++
m_rect = wxRect(m_window->GetScreenPosition(), dc->GetSize());
```

I think, without any dynamic casts.

AliKet replied:
When `wxClientDC` is used we want the overlay window to be positioned over the client area of the window. `m_window->GetScreenPosition()` returns the upper left corner of the window and not the top left position of the client area.

The overlay window positioned at `m_window->GetScreenPosition()` (**incorrect**): 

![overlay_incorrect](https://github.com/wxWidgets/wxWidgets/assets/7704771/d5b04847-3eab-4bb1-ad52-1c0c8080d29c)

The overlay window positioned at `m_window->GetClientRect()` (**correct**):

![overlay_correct](https://github.com/wxWidgets/wxWidgets/assets/7704771/18e80882-2fc2-4e26-b797-6a899be00759)

vadz replied:
Ah, I see, thanks for the explanation!

I'd still like to avoid dynamic cast, for all the usual reasons. As always, we can just add a virtual function instead and while it's a bit ugly to have window-oriented function in `wxDC`, we do already have `GetWindow()` there, so we could also add some `GetOriginInWindow()` or `GetWindowOffset()` or whatever else you'd prefer that would return `wxPoint(0, 0)` by default but would be overridden to return `GetClientAreaOrigin()` in `wxClientDCImpl`.

Do you see any problems with doing it like this? Thanks again!

vadz commented on src/common/overlaycmn.cpp:
```diff
@@ -140,6 +140,20 @@ void wxDCOverlay::Clear()
 
 #include "wx/window.h"
 
+#if defined(__WXGTK__) && !defined(__WXGTK3__)
```
I would just put this in `Init()` itself, i.e. avoid defining the extra function which is used for wxGTK2 only but just inlined its body there inside `#if GTK2` test.

The rationale is that it's not obvious when seeing the call to `wxDontClipSubWindows()` that it only does something for GTK2, and it would be helpful to see it immediately.

This would also be a bit shorter.

vadz commented on src/generic/splitter.cpp:
```diff
@@ -71,17 +71,37 @@ wxEND_EVENT_TABLE()
+    return wnd->HasFlag(wxSP_LIVE_UPDATE);
+}
+
+static wxBitmap wxPaneCreateStippleBitmap()
```
We should extract this function somewhere and probably cache the bitmap because even if it's just a 2x2 one, I am still not sure we want to recreate it on every repaint.

The trouble is that we need to clean it up in some module and I don't see anything obviously appropriate. Perhaps we could make this a stock brush? Then it would be cleaned up by `wxStockGDI::DeleteAll()`.

But at least let's factor out this function into some private header, it's not quite trivial and it might need to be changed later (for dark mode or high DPI or I don't know what else) and so I'd rather not have 2 copies of it.

AliKet replied:
Updating the documentation and this refactoring are still not done yet. if you don't mind me asking you to do that when testing under wxOSX, or fixing any regressions introduced (which I haven't noticed any!). TIA!

AliKet replied:
Would it be better to do the refactoring in a seperate PR instead?

vadz reviewed (commented):
Thanks a lot, this is much better and I'd really like to merge it but I'd like to avoid duplicating the function for creating the stipple bitmap. Will you be able to change it or should I do it myself?

vadz commented:
> I also managed to get transparent hint to work with `wxOverlay` (generic and native), so `wxPseudoTransparentFrame` (with all its complications and limitation) is no longer needed.

Excellent!

> I just want to know if it's ok to push the changes (depend on the changes done here) here or should I wait until this is approved and merged?

Whatever you prefer. I.e. I'm fine with adding them here.

> as always I need someone else to test my changes under **wxOSX**. TIA!

I'll do it before merging, so from this point of view it would be better to remove `wxPseudoTransparentFrame` now, as I could test this too then.

Thanks again!

AliKet committed:
Use wxOverlay to show docking hint instead of transparent wxFrame

The main advantage of using wxOverlay over wxFrame, besides reduced
code complexity, is that the docking hint now works correctly under
Wayland which it didn't work before.

wxAUI_MGR_RECTANGLE_HINT now works everywhere (including wxGTK and wxOSX)

We simply use a bitmap to draw the venetian blinds hint instead of the
exotic and esoteric code using wxRegion.

wxGTK2 now uses graphics context for transparency. If not available, the hint
will be drawn using wxINVERT raster operation mode.

vadz reviewed (commented):
Thanks again for all your work here, this looks really nice but I just have a couple of questions about opacity semantics under MSW, notably its default behaviour, could you please check this? TIA!

AliKet commented on src/msw/overlay.cpp:
```diff
@@ -100,12 +125,21 @@ void wxOverlayImpl::Init(wxDC* dc, int , int , int , int )
+    }
 
-    m_overlayWindow = wxCreateOverlayWindow(m_rect);
+    HWND hWndInsertAfter = ::GetNextWindow(GetHwndOf(m_window), GW_HWNDPREV);
```
This change is mainly to prevent the overlay window from being placed on top of/covering any floating window.

**Before:**

![overlay_bad](https://github.com/wxWidgets/wxWidgets/assets/7704771/8e01671b-9f6e-4bc6-aabf-49f968f5b5a3)

**After:**

![overlay_good](https://github.com/wxWidgets/wxWidgets/assets/7704771/2bd98289-8629-4cbc-aace-7fe1f6a5ea05)

Unfortunately, this breaks rubberbanding in the drawing sample. That is, the rubberband no longer appears there! What actually solves both problems is doing the following:
```
    const auto win = wxGetTopLevelParent(m_window);
    HWND hWndInsertAfter = ::GetNextWindow(GetHwndOf(win), GW_HWNDPREV);
    m_overlayWindow = wxCreateOverlayWindow(m_rect, m_alpha, hWndInsertAfter);
```
Which I'll commit if you don't see any problem in doing so (of course) ?

AliKet commented on include/wx/dc.h:
```diff
@@ -713,6 +719,9 @@ class WXDLLIMPEXP_CORE wxDC : public wxObject
     wxWindow *GetWindow() const
         { return m_pimpl->GetWindow(); }
 
+    wxRect GetWindowRect() const
```
Documented now!

AliKet committed:
wxAuiManager::AlwaysUsesLiveResize() documentation updated

This is no longer true as wxAUI now uses wxOverlay to display the resizing hint
in non-live resize mode, which should work on all platforms.

AliKet commented on src/aui/framemanager.cpp:
```diff
@@ -744,40 +559,23 @@ wxAuiDockUIPart* wxAuiManager::HitTest(int x, int y)
 }
 
-/* static */ bool wxAuiManager::AlwaysUsesLiveResize(const wxWindow* window)
+/* static */ bool wxAuiManager::AlwaysUsesLiveResize(const wxWindow* WXUNUSED(window))
```
Updated.

AliKet commented:
@vadz I see that [`wxSplitterWindow::AlwaysUsesLiveUpdate`](https://docs.wxwidgets.org/latest/classwx_splitter_window.html#a7a377a089bb00011d00f233b055a5f8f)  is only added in 3.3.0.
Should we remove it completely or update its documentation like we did with `wxAuiManager::AlwaysUsesLiveResize`?

vadz commented:
> @vadz I see that [`wxSplitterWindow::AlwaysUsesLiveUpdate`](https://docs.wxwidgets.org/latest/classwx_splitter_window.html#a7a377a089bb00011d00f233b055a5f8f) is only added in 3.3.0. Should we remove it completely or update its documentation like we did with `wxAuiManager::AlwaysUsesLiveResize`?

Yes, I think it should be fine to remove it if we never do this any longer. Thanks!

paulcor commented on interface/wx/overlay.h:
```diff
@@ -27,6 +27,25 @@ class wxOverlay
+        If @a alpha is `-1`, use per-pixel alpha blending, otherwise use the
+        given alpha value for all pixels.
+
+        @note The overlay window is opaque by default under wxMSW. You have to
```
That's not my understanding of how 'wxOverlay` is supposed to work. It's expected that calling `wxDCOverlay::Clear()` will make the overlay transparent.

AliKet replied:
In fact, it works as you expect. Just the wording is misleading and should say something like this instead:
> Drawing on the overlay window is opaque by default under wxMSW...

because this is how `SetLayeredWindowAttributes` works. i.e. it can only apply a constant alpha on the drawing with no alpha blending at all.

As pictures speak louder than words, this is what I mean with rubberbanding as example:

- Rubberbanding (outline) with `SetLayeredWindowAttributes` + `wxALPHA_OPAQUE` (default):

![overlay_opaque_outline](https://github.com/wxWidgets/wxWidgets/assets/7704771/21b6dbeb-05b5-4f09-9113-5894c7cb868d)

- Rubberbanding (filled) with `SetLayeredWindowAttributes` + `wxALPHA_OPAQUE` :

![overlay_opaque](https://github.com/wxWidgets/wxWidgets/assets/7704771/c2400b2a-3740-4ad3-b9c8-e06b93caae7a)

- Rubberbanding (filled) with `SetLayeredWindowAttributes` + alpha (128): 

![overlay_translucent](https://github.com/wxWidgets/wxWidgets/assets/7704771/8022d6c4-ee9b-4314-8154-f092f6f80c9d)

- Rubberbanding (filled) with `UpdateLayeredWindow` + wxGCDC: 

![overlay_translucent_gc](https://github.com/wxWidgets/wxWidgets/assets/7704771/8b3bd91e-eb40-4bfe-b3dd-689ea1dc7417)

AliKet commented on include/wx/dc.h:
```diff
@@ -176,6 +176,12 @@ class WXDLLIMPEXP_CORE wxDCImpl: public wxObject
+    // coordinates as returned by GetScreenRect(). It returns the client
+    // rectangle (also in screen coordinates) if the DC is a wxClientDC.
+    // This is mostly meant to be used by wxOverlay.
+    virtual wxRect GetWindowRect() const;
```
I'll remove this commit then.

BTW, `wxOverlay` and `wxDCOverlay` documentations need to be updated. Also mentione that `wxClientDC` is the only supported DC that should be used to draw on the overlay!

Hope I understood you correctly!!

AliKet committed:
Fix wxOverlay when used with a TLW under wxMSW

Correctly position the overlay window over the TLW's client area.

AliKet committed:
Added per-pixel alpha blending capability to wxOverlay under wxMSW

By default, wxMSW overlay uses a constant alpha for the window's opacity which
can be changed by the added wxOverlay::SetOpacity() function. For some applications,
that's all they need. But for others, per-pixel alpha blending is a requirement.
This capability can be enabled by calling SetOpacity(-1) before initializing the
overlay. But the drawing must be done via a graphics context because the standard DC
under MSW does not support alpha drawing.

Also request that the overlay window be positioned below any floating windows
on the target (if any).

vadz committed:
Avoid using "x2" and "y2" for sash width and height

This was confusing, so get rid of these variables and use more clearly
named "posSash" and "sizeSash" instead.

No real changes.

vadz committed:
Inline DrawResizeHint() function into its only caller

It seems to be more clear to just draw the sash in DrawSashTracker()
itself instead of calling a function which does it.

vadz committed:
Get rid of IsLive() helper in wxSplitterWindow code

This was useful when we had AlwaysUsesLiveResize() and the related
complications, but now that live resizing is governed solely by
wxSP_LIVE_UPDATE again, it's not needed any longer.

No real changes.

vadz committed:
Initialize wxOverlayImpl members in their declarations

Ensure that all member variables are initialized instead of initializing
just some of them in the ctor and leaving others uninitialized, it's
simpler to just do this rather than to check if uninitialized values can
or not be used.

vadz committed:
Erase overlay contents when resetting it in wxOSX

Without this, the old contents continues to be shown for some time
before the display was refreshed.

vadz reviewed (approved):
Thanks a lot once again for all your work on this!

I've finally tested this under Mac and I found one problem there: the overlay contents continued to be displayed after `Reset()` (although it disappeared as soon as the window was redrawn). I couldn't find any better way to fix this than the brute force approach implemented in 4261e1d07d52c0703a5adf4a436937eb0f8896e2 but at least with this it works for me.

@csomor If you have a moment to look at this commit and let me know if there is some simpler way of removing the overlay contents from screen, please let me know. FWIW I've tried calling `setNeedsDisplay:YES` but this didn't seem to help.

vadz committed:
Avoid implicit "this" capture in a lambda to avoid C++20 warning

Implicitly capturing "this" is deprecated in C++20, so do it explicitly,
even if it also requires explicitly capturing everything else as
combining implicit capture with explicitly capturing "this" is not
allowed in the previous C++ versions.

vadz commented:
I'm merging this as it has been opened for a long time. We can almost surely improve things under Mac, but this can be done later.

Thanks again for all your efforts!
